common/lib/plugins/bluegreen/blue_green_plugin.ts

```diff
@@ -0,0 +1,236 @@
+      let routing: ConnectRouting | undefined = this.bgStatus.connectRouting.filter((routing: ConnectRouting) =>
+        routing.isMatch(hostInfo, hostRole)
+      )[0];
+
+      if (!routing) {
+        return this.regularOpenConnection(connectFunc, isInitialConnection);
+      }
+
+      this.startTimeNano = getTimeInNanos();
+      while (routing && !client) {
+        client = await routing.apply(this, hostInfo, props, isInitialConnection, connectFunc, this.pluginService);
+        if (!client) {
+          this.bgStatus = this.pluginService.getStatus<BlueGreenStatus>(BlueGreenStatus, this.bgdId);
+          if (! this.bgStatus) {
+            this.endTimeNano = getTimeInNanos();
+            return this.regularOpenConnection(connectFunc, isInitialConnection);
+          }
+          routing = this.bgStatus.connectRouting.filter((routing: ConnectRouting) => routing.isMatch(hostInfo, hostRole))[0];
+        }
+      }
+
+      this.endTimeNano = getTimeInNanos();
+      if (!client) {
+        client = await connectFunc();
+      }
+
+      if (isInitialConnection) {
+        // Provider should be initialized after connection is open and a dialect is properly identified.
+        this.initProvider();
+      }
+
+      return client;
+    } finally {
+      if (this.startTimeNano > 0 && this.endTimeNano === BigInt(0)) {
+        this.endTimeNano = getTimeInNanos();
+      }
+    }
+  }
+
+  async execute<T>(methodName: string, methodFunc: () => Promise<T>, methodArgs: any[]): Promise<T> {
+    this.resetRoutingTimeNano();
+
+    try {
+      this.initProvider();
+
+      if (BlueGreenPlugin.CLOSED_METHOD_NAMES.has(methodName)) {
+        return await methodFunc();
+      }
+
+      this.bgStatus = this.pluginService.getStatus<BlueGreenStatus>(BlueGreenStatus, this.bgdId);
+
+      if (!this.bgStatus) {
+        this.endTimeNano = getTimeInNanos();
+        return await methodFunc();
+      }
+
+      const currentHostInfo: HostInfo = this.pluginService.getCurrentHostInfo();
+      const hostRole: BlueGreenRole = this.bgStatus.getRole(currentHostInfo);
+
+      if (!hostRole) {
+        // Connection to a host that isn't participating in BG switchover.
+        return await methodFunc();
+      }
+
+      let result: T | null = null;
+      let routing: ExecuteRouting | undefined = this.bgStatus.executeRouting.filter((routing: ExecuteRouting) =>
+        routing.isMatch(currentHostInfo, hostRole)
+      )[0];
+
+      if (!routing) {
+        return await methodFunc();
+      }
+
+      this.startTimeNano = getTimeInNanos();
+
+      while (routing && !result) {
+        result = await routing.apply(this, methodName, methodFunc, methodArgs, this.properties, this.pluginService);
+        if (!result) {
+          this.bgStatus = this.pluginService.getStatus<BlueGreenStatus>(BlueGreenStatus, this.bgdId);
+          routing = this.bgStatus.executeRouting.filter((routing: ExecuteRouting) => routing.isMatch(currentHostInfo, hostRole))[0];
```

Comment:
The bug I asked you to port over from JDBC for connect() is similar to another bug fix in that same PR that needs to be ported over as well, if the status is null we should set the end time and then call the methodFunc